flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java

```diff
@@ -1058,6 +1071,21 @@ public CompletableFuture<CoordinationResponse> deliverCoordinationRequestToCoord
                 operator, request);
     }
 
+    @Override
+    public boolean notifyTaskFinishing(ExecutionAttemptID executionAttemptID) {
+        BitSet subtaskStatus = finishingTasks.get(executionAttemptID.getJobVertexId());
+        subtaskStatus.set(executionAttemptID.getSubtaskIndex());
+
+        for (Map.Entry<JobVertexID, BitSet> entry : finishingTasks.entrySet()) {
+            JobVertexID vertex = entry.getKey();
+            BitSet status = entry.getValue();
+            if (status.cardinality() != executionGraph.getJobVertex(vertex).getParallelism()) {
```

Comment:
Can we avoid having to check every vertex each time a subtask reaches end-of-data?

For example, maybe we can rename  `finishingTasks` to `tasksWaitingEndOfData` and remove job vertex every time all subtasks of a job vertex has reached end of data. Then trigger checkpoint when the map is empty.